app.js

```diff
@@ -0,0 +1,157 @@
+//jshint esversion: 6
+const express = require("express");
+const bodyParser = require("body-parser");
+const mongoose = require("mongoose");
+const redisClient = require("redis").createClient;
+const redis = redisClient(6379, 'redis-server');
+
+const app = express();
+
+app.set('view engine', 'ejs');
+app.use(bodyParser.urlencoded({
+  extended: true
+}));
+app.use(express.static("public"));
+//let studentMap = new Map();
+
+mongoose.connect("mongodb://mongo:27017/mongorediscomposeDB", {
+  useNewUrlParser: true,
+  useUnifiedTopology: true,
+  useFindAndModify: false
+}, function(err) {
+  if (!err) {
+    console.log("successfully connected");
+  }
+});
+
+redis.on("connect", function() {
+  console.log('connected to Redis');
+});
+
+app.listen(3000, function() {
+  console.log("Server started at https://localhost:3000");
+});
+
+const studentSchema = mongoose.Schema({
+  name: String,
+  roll: Number,
+  physicsMarks: Number,
+  chemistryMarks: Number,
+  mathsMarks: Number
+});
+
+const Student = mongoose.model("Student", studentSchema);
+
+//console.log(studentMap);
+
+app.route("/students")
```

Comment:
This a very clear example of callback hell

Try to read about it, if you are not familiar with the concept there are other ways to write the same code which will be structurally better and it won't just keep going to the left of the file and also there will be lesser scopes.

You might already be aware about what a scope means, but just for the explanation here let me give a brief description here. So a for loop, an if else or a function defines a scope. Usually anything that comes within a curly braces (with an exception of dictionaries).

Then you have nested scopes where you have an if else within a loop which is within a loop which may again be within an if else and so on..., this creates more [Cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity). More the complexity, will make it more difficult for the program to read and to implement test cases over it.

In newer versions of node, we have promises and also async, await to deal with promises more efficiently. They may not be very easy to understand in first go but helps a lot and helps in writing a very clean code.